File: src/models/Individual.ts
import Town from "./Town";

export default class Individual {
    public towns: Town[];

    constructor(towns: Town[]) {
        this.towns = [...towns];
    }

    public shuffle() {
        for (let i = this.towns.length - 1; i > 0; i--) {
            const j = Math.floor(Math.random() * (i + 1));
            [this.towns[i], this.towns[j]] = [this.towns[j], this.towns[i]];
        }
    }

    public makeGreedy(startingTown: Town) {
        const visitedTowns = new Set<Town>();
        visitedTowns.add(startingTown);
        let currentTown = startingTown;

        while (visitedTowns.size < this.towns.length) {
            let minDistance = Number.MAX_VALUE;
            let minTown;

            for (let town of this.towns) {
                if (visitedTowns.has(town)) continue;

                const distance = currentTown.calculateDistance(town);
                if (distance < minDistance) {
                    minDistance = distance;
                    minTown = town;
                }
            }

            if (!minTown) throw new Error("Haven't found the closest town");

            visitedTowns.add(minTown);
            currentTown = minTown;
        }

        this.towns = Array.from(visitedTowns);
    }

    public calculateTotalDistance() {
        let sum = 0;

        for (let i = 0; i < this.towns.length - 1; i++)
            sum += this.towns[i].calculateDistance(this.towns[i + 1]);

        sum += this.towns[0].calculateDistance(
            this.towns[this.towns.length - 1]
        );

        return Math.round((sum + Number.EPSILON) * 100) / 100;
    }
}

File: zad1/Models/Individual.cs
using System;
using System.Collections.Generic;

namespace zad1.Models
{
    public class Individual : ICloneable, IComparable
    {
        public List<Town> Towns { get; set; }

        private double? TotalDistance { get; set; }

        public Individual(IEnumerable<Town> towns)
        {
            Towns = new List<Town>(towns);
        }

        public Individual Shuffle()
        {
            var rng = new Random();

            for (var i = Towns.Count - 1; i > 0; i--)
            {
                var j = rng.Next(i + 1);

                var temp = Towns[i];
                Towns[i] = Towns[j];
                Towns[j] = temp;
            }

            return this;
        }

        public Individual MakeGreedy(Town startingTown)
        {
            var visitedTowns = new HashSet<Town> {startingTown};
            var currentTown = startingTown;

            while (visitedTowns.Count < Towns.Count)
            {
                var minDistance = double.MaxValue;
                Town minTown = null;

                foreach (var town in Towns)
                {
                    if (visitedTowns.Contains(town)) continue;
                    var distance = currentTown.CalculateDistance(town);

                    if (distance < minDistance)
                    {
                        minDistance = distance;
                        minTown = town;
                    }
                }

                if (minTown is null) throw new Exception("Haven't found the closest town");

                visitedTowns.Add(minTown);
                currentTown = minTown;
            }

            Towns = new List<Town>(visitedTowns);
            return this;
        }

        public double CalculateTotalDistance()
        {
            if (TotalDistance.HasValue) return TotalDistance.Value;

            var sum = Towns[0].CalculateDistance(Towns[^1]);

            for (var i = 0; i < Towns.Count - 1; i++)
                sum += Towns[i].CalculateDistance(Towns[i + 1]);

            TotalDistance = Math.Round(sum, 2);

            return TotalDistance.Value;
        }

        public object Clone()
        {
            return MemberwiseClone();
        }

        public int CompareTo(object? obj)
        {
            if (obj is Individual otherIndividual)
                return CalculateTotalDistance().CompareTo(otherIndividual.CalculateTotalDistance());
            return 1;
        }
    }
}